### ISSUE TITLE
Relatively high idle CPU usage reported by about:performance, increases with number of addons

### ISSUE THREAD

**Issue by nkestrel** [JasonBarnabe/stylish#275]
Developer Edition (47.0a2 2016-03-12) just got about:performance for comparing addon resource usage and I noticed Stylish was always at the top of the list. With no styles installed and the browser idling Stylish uses 11% CPU and 2% System while all my other addons use less than 1% CPU and 0% System.

I tried a clean profile with only Stylish installed, CPU usage was lower (4% CPU and 1% System) but installing more addons steadily increased it until it matched my main profile.

**Comment by easyaspi314** [comment]
+1 confirmed.

**Comment by SW1FT** [comment]
I have around 3 styles and I'm seeing 3~6% CPU usage and 0~1% System usage...

**Comment by easyaspi314** [comment]
@SW1FT are you using Developer Edition?

**Comment by SW1FT** [comment]
Yes, I'm using 2 global styles for FF's UI and the other one is site-specific.

**Comment by easyaspi314** [comment]
And you are using Developer Edition or Nightly with `e10s`, right? 
Hover over a tab on a website (this site, for example) and it should say, "e10s" on it. If you are using Dev Edition and it doesn't say that, go to Preferences and make sure multiprocess is checked.

**Comment by SW1FT** [comment]
I'm not a retard, I know I'm using e10s and on FF DE.

**Comment by easyaspi314** [comment]
Jesus Christ! I never implied that!

**Comment by johndrinkwater** [comment]
While debugging a stall in firefox, I arrived here and wanted to highlight that while using `about:performance`, stylish.sqlite is getting opened once for each of the visible tabs shown, and repeats for each timer, so it may be that the reporting system is triggering an abnormal code path

```
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60
open("~/.mozilla/firefox/johndrinkwater/stylish.sqlite", O_RDWR|O_CREAT|O_CLOEXEC, 0644) = 60  
```

That’s not to say Stylish is doing something incorrectly or not. Just needs more looking into.

**Comment by easyaspi314** [comment]
`stylish.sqlite` is where Stylish stores its styles.

**Comment by Gitoffthelawn** [comment]
@easyaspi314 I'm guessing @johndrinkwater is aware that `stylish.sqlite` is where Stylish stores its styles.  He's observing that it might be getting opened excessively, thus impacting browser performance.

**Comment by easyaspi314** [comment]
@Gitoffthelawn 

Yeah. I know. I am new to this git. And maybe, what it is doing is opening that file in the content process instead of the chrome process. I dunno, I literally know nothing about the api. I know basic HTML and CSS; and a bit of jQuery. That is it.

[![e10s-overview](https://cloud.githubusercontent.com/assets/6258309/15451055/ed1596d6-1f7e-11e6-97b7-9761b61aa3ca.png)](https://developer.mozilla.org/en-US/Add-ons/Working_with_multiprocess_Firefox)

Are we shimming or running in native mode?

**Comment by nkestrel** [comment]
When I reported this problem I recall it happening without `about:performance` being open, there was a small continuous idle CPU usage in the system monitor that could be eliminated by disabling Stylish.

I tried to reproduce this just now with 48.0a2 (2016-05-21) (same Stylish 2.0.6 without any styles installed) and the situation has changed, `about:performance` now reports Stylish using 2% CPU and 1% System which is significantly lower than before despite running on the same system with similar number of addons and tabs. CPU usage no longer appears to increase with the number of addons. Performance analysis shows Stylish doing work at 5 second intervals when `about:performance` is open but not when it is closed. 

The original problem appears to have mostly resolved itself and is now just limited to the `about:performance` page.

**Comment by Gitoffthelawn** [comment]
@nkestrel What's it doing at 5 second intervals?

**Comment by nkestrel** [comment]
@Gitoffthelawn `findSql()` mostly, which matches what @johndrinkwater reported.

Call stack:
```
Style.prototype.findSql
Style.prototype.find
UserStyleManager.getAddonByID
AddonManager.getAddonByID
```

**Comment by Gitoffthelawn** [comment]
Using about:performance on FF 47, I'm seeing 13% CPU usage.  That's 550% more than the next highest extension.

Also seeing more *high-impact*, *medium-impact*, and *alert* framerate events than any other extension.

**Comment by JasonBarnabe** [comment]
Stylish registers the styles you've installed as addons in Firefox so they show up in the addons manager. about:performance asks for a list of addons every 5 seconds. So about:performance is causing the CPU usage it's reporting.

I don't know how about:performance is doing what it's doing, but potentially it could either not ask for the types of addons it's not going to report on (maybe it does and Stylish isn't listening?), or stop recording performance info when it itself is refreshing.

**Comment by Gitoffthelawn** [comment]
That's a great explanation.  Thanks Jason.

Would you be willing to communicate with Mozilla (via bugzilla or otherwise) so they could look into this issue.  Your extension is one of the most popular in the world, so they might be willing to improve how they generate their metrics.

Also, does your explanation also satisfy the report by @johndrinkwater , or is that something different?

Thanks again!

**Comment by JasonBarnabe** [comment]
I don't think it's really worth it - if it's just doing this when about:performance is open, who cares, really. In any case, the current plan is to toss all of this and use the code from Stylish for Chrome.

**Comment by Gitoffthelawn** [comment]
@JasonBarnabe Good point.  Except if there is a Stylish performance issue that is being masked by how *about:performance* is coded, we might never know!  That might still be true when you switch to WebExtensions.

Does your explanation satisfy the report by @johndrinkwater , or is that something different?

**Comment by edwardgalligan** [comment]
> if there is a Stylish performance issue that is being masked by how about:performance is coded, we might never know!

Just to add an anecdata-point to this, I installed stylish recently, saw my laptop (MB Pro) get loud and hot, and battery life dropped off. After a few days of this, I tried investigating: Activity Monitor pointed me to Firefox, `about:performance` pointed me to Stylish, and uninstalling the add-on restored normality.

I can't see how the above would be caused by the `about:performance` tool, as I only used it once, quite briefly.
